File: src/ca/uqam/mgl7460/tp1/types/traitements/instances/InstanceProcessus.java
package ca.uqam.mgl7460.tp1.types.traitements.instances;

import java.time.Instant;
import java.util.logging.Logger;

import ca.uqam.mgl7460.tp1.types.modeles.DemandePret;
import ca.uqam.mgl7460.tp1.types.traitements.definitions.DefinitionProcessus;

/**
 * Ce type représente une exécution particulière d'un modèle de processus (
 * <code>DefinitionProcessus</code>).
 * 
 * Cette exécution est caractérisée par:
 * 1) la definitionProcessus (modèle de processus) correspondant-e
 * 2) la demande de pret faisant l'pobjet de l'exécution
 * 3) le temps de démarrage
 * 4) le temps d'arrêt
 * et à tout moment:
 * 5) l'état actuel d'exécution du processus
 * 6) la tâche courante
 * 
 * Concernant l'exécution, c'est (évidemment) l'instance de processus qui "mène".
 * 
 * La méthode <code>void demarrer() throws ExceptionDefinitionProcessus</code> lance
 * l'exécution en allant chercher la première <code>DefinitionTache</code> de la
 * <code>definitionProcessus</code>, l'instancie (créer une <code>InstanceTache</code>)
 * et lance l'exécution de code>InstanceTache</code>.
 * 
 * Quand l'instance termine son exécution, elle "enregistre ses résultats" sur la demande de
 * prêt (demandePret.setResultatTraitement(...)) et 'crie "j'ai fini" ' auquel case l'instance 
 * de processus consulte le modèle de processus (getDefinitionProcessus()) pour voir, selon
 * le modèle, qu'elle est la tâche suivante.
 * 
 * Le 'cri' "j'ai fini", c'est en fait l'appel, par les instances de tâches quand elles ont fini,
 * de la méthode <code>void signalerFinTache(InstanceTache instanceTache) throws ExceptionDefinitionProcessus</code>
 * 
 * Donc, c'est la fonction <code>void signalerFinTache(InstanceTache instanceTache) throws ExceptionDefinitionProcessus</code>
 * qui va identifier la (définition de) tache suivante à instancier et à démarrer.
 * 
 * Si elle n'en trouve pas, c'est la fin du processus
 * 
 * Si elle en trouve plusieurs (prochaines taches candidates), c'est que le processus est mal formé, et c'est dans ce cas
 * que la fonction  <code>void signalerFinTache(...) throws ExceptionDefinitionProcessus</code> lance
 * l'exception <code>ExceptionDefinitionProcessus</code>
 */
public interface InstanceProcessus {

    /**
     * retourne la demande de pret faisant l'objet de l'exécution
     * @return
     */
    public DemandePret getDemanderPret();

    /**
     * retourne la définition (modèle) de processus, dont cette exécution
     * est une instance
     * @return
     */
    public DefinitionProcessus getDefinitionProcessus();

    /**
     * retourne l'état actuel de l'exécution du processus. Ça consiste
     * en deux choses:
     * 1)   La tâche courante (on peut y penser comme un pointeur/curseur
     *      sur la tâche qui est sur le point de démarrer, en train d'exécuter,
     *      ou qui vient de se terminer)
     * 2)   l'état d'exécution (<code>EtatTraitement</code>) qui peut etre l'un de
     *      PRET, ENCOUR ou TERMINE
     * @return
     */
    public EtatProcessus getEtatProcessus();

    /**
     * modifie l'état du processus
     * @param etatProcessus
     */
    public void setEtatProcessus(EtatProcessus etatProcessus);

    /**
     * retourne la tache courante
     * @return
     */
    public InstanceTache getTacheCourante();

    /**
     * modifie la tâche courante
     * @param tacheCourante
     */
    public void setTacheCourante(InstanceTache tacheCourante);

    /**
     * cette méthode va instancier la première tâche de la <code>definitionProcessus</code>
     * et la lancer (l'exécuter). Quand la première instance de tache aura fini, elle appelera 
     * la méthode <code>void signalerFinTache(InstanceTache instanceTache) throws ExceptionDefinitionProcessus</code>
     * sur le processus. Voir suite de l'explication dans la documentation de l'entête de l'interface
     * @throws ExceptionDefinitionProcessus
     */
    public void demarrer() throws ExceptionDefinitionProcessus;

    /**
     * retourne le temps de démarrage
     * @return
     */
    public Instant getTempsDemarrage();

    /**
     * initialise (modifie) le temps de démarrage
     */
    public void setTempsDemarrage(Instant temps);

    /**
     * retourne le temps d'arret
     * @return
     */
    public Instant getTempsArret();

    /**
     * initialise (modifie) le temps d'arret
     * @param temps
     */
    public void setTempsArret(Instant temps);

    /**
     * une méthode appelée par les <code>InstanceTache</code> quand elles font
     * fini d'exécuter. Le récepteur (instance de processus) va alors consulter la 
     * définition (le modèle) de processus qu'il est entrain d'exécuter pour identifier
     * la prochaine tâche à exécuter. Si la méthode n'en trouve pas, cela marquera la
     * fin de l'exécution du processus.
     * 
     * Si on en trouve plusieurs, problème! le processus est mal formé, ce qui va enclencher
     * l'exception <code>ExceptionDefinitionProcessus</code>
     * @param instanceTache
     * @throws ExceptionDefinitionProcessus
     */
    public void signalerFinTache(InstanceTache instanceTache) throws ExceptionDefinitionProcessus;

    /**
     * retourne le <code>java.util.logging.Logger</code> utiliser par le processus pour le "tracing",
     * au besoin.
     * @return
     */
    public Logger getLogger();

    /**
     * modifie le logger
     * @param logger
     */
    public void setLogger(Logger logger);
    
}

File: src_python/traitements/instances/instance_processus.py

import datetime
import logging
from traitements.instances.etat_processus import EtatProcessus
from traitements.instances.etat_traitement import EtatTraitement
from traitements.instances.exception_definition_processus import ExceptionDefinitionProcessus
from traitements.instances.instance_tache import InstanceTache


class InstanceProcessus:
    def __init__(self, definition_processus, demande_pret):
        self.definition_processus = definition_processus
        self.demande_pret = demande_pret

        self.taches = []
        self.etat_processus: EtatProcessus = None
        self.tache_courante: InstanceTache = None

        self.temps_demarrage = None
        self.temps_arret = None
        self.logger = None
        
    
    @property
    def etat_processus(self):
        return self._etat_processus
    @etat_processus.setter
    def etat_processus(self, value):
        self._etat_processus = value

    @property
    def tache_courante(self):
        return self._tache_courante
    @tache_courante.setter
    def tache_courante(self, value):
        self._tache_courante = value
    
    @property
    def temps_demarrage(self):
        return self._temps_demarrage
    @temps_demarrage.setter
    def temps_demarrage(self, value: datetime):
        self._temps_demarrage = value
    
    @property
    def temps_arret(self):
        return self._temps_arret    
    @temps_arret.setter
    def temps_arret(self, value: datetime):
        self._temps_arret = value

    @property
    def logger(self):
        return self._logger
    @logger.setter
    def logger(self, value: logging.Logger):
        self._logger = value

    def demarrer(self):
        premiere_tache = self.definition_processus.premiere_tache
        if (premiere_tache is None):
            raise ExceptionDefinitionProcessus(self, [], "Le processus n'a pas de première tâche définie.")
        premiere_instance_tache = InstanceTache(self, premiere_tache)
        self.temps_demarrage = datetime.datetime.now()
        self.ajouter_et_lancer_tache(premiere_instance_tache)

    
    def signaler_fin_tache(self, instance_tache: InstanceTache):
        if (instance_tache != self.tache_courante):
            raise ExceptionDefinitionProcessus(self, [], "La tâche signalée comme terminée n'est pas la tâche courante.")
        
        if (instance_tache.etat_instance_tache != EtatTraitement.TERMINE):
            raise ExceptionDefinitionProcessus(self, [], "La tâche signalée comme terminée n'est pas dans un état terminé.")
        
        transitions_sortantes = self.definition_processus.get_trasitions_sortant_de(instance_tache.definition_tache)

        if (len(transitions_sortantes) == 0):
            self.temps_arret = datetime.datetime.now()
            self.etat_processus = EtatProcessus(instance_tache, EtatTraitement.TERMINE)
            return
        elif (len(transitions_sortantes) > 1):
            raise ExceptionDefinitionProcessus(self, [], f"La tâche {instance_tache.definition_tache.nom} signalant la fin a plus d'une transition sortante.")
        else:
            prochaineTransition = transitions_sortantes[0]
            prochaineInstanceTache = InstanceTache(self, prochaineTransition.tache_destination)
            self.ajouter_et_lancer_tache(prochaineInstanceTache)

    def ajouter_et_lancer_tache(self, instanceTache: InstanceTache):
        self.etat_processus = EtatProcessus(instanceTache, EtatTraitement.PRET)
        self.taches.append(instanceTache)
        self.tache_courante = instanceTache
        instanceTache.executer()
        self.etat_processus = EtatProcessus(instanceTache, EtatTraitement.ENCOURS)
        


    